contracts/core/extensions/CoreIssuanceOrder.sol

```diff
@@ -36,17 +36,131 @@ contract CoreIssuanceOrder is
 {
     using SafeMath for uint256;
 
+    string constant INVALID_SIGNATURE = "Invalid order signature.";
+
+    struct IssuanceOrder {
+        address setAddress;
+        uint256 quantity;
+        address makerAddress;
+        address makerToken;
+        uint256 makerTokenAmount;
+        uint256 expiration;
+        address relayerToken;
+        uint256 relayerTokenAmount;
+        uint256 salt;
+        bytes32 orderHash;
+    }
+
     function fillOrder(
-        address _maker,
-        address _setAddress,
-        uint _quantity
+        address[4] _addresses,
+        uint[5] _values,
+        uint _fillQuantity,
+        uint8 _v,
+        bytes32 _r,
+        bytes32 _s
     )
         public
-        isValidSet(_setAddress)
-        isPositiveQuantity(_quantity)
-        isNaturalUnitMultiple(_quantity, _setAddress)
+        isValidSet(_addresses[0])
+        isPositiveQuantity(_fillQuantity)
+        isNaturalUnitMultiple(_fillQuantity, _addresses[0])
     {
+
+        IssuanceOrder memory order = IssuanceOrder({
+            setAddress: _addresses[0],
+            quantity: _values[0],
+            makerAddress: _addresses[1],
+            makerToken: _addresses[2],
+            makerTokenAmount: _values[1],
+            expiration: _values[2],
+            relayerToken: _addresses[3],
+            relayerTokenAmount: _values[3],
+            salt: _values[4],
+            orderHash: generateOrderHash(
+                _addresses,
+                _values
+            )
+        });
+
+        // Verify order is valid
+        validateOrder(
+            order,
+            _fillQuantity
+        );
+
+        // Verify signature is authentic
+        require(
+            validateSignature(
+                order.orderHash,
+                order.makerAddress,
+                _v,
+                _r,
+                _s
+            ),
+            INVALID_SIGNATURE
+        );
+
         //Issue Set
-        issueInternal(_maker, _setAddress, _quantity);
+        issueInternal(
+            order.makerAddress,
+            order.setAddress,
+            _fillQuantity
+        );
+    }
+
+    function generateOrderHash(
+        address[4] _addresses,
+        uint[5] _values
+    )
+        public
```

Comment:
Public? This seems like something that should be internal or private

Reply:
Yeah this one I struggled with, there's no harm in making it public but makes more sense for it to be internal. For expediency purposes I just made it public so that it could be tested. We haven't come up with a framework for testing internal functions I feel like.

Reply:
I think these belong in a library of some sort as they don’t alter state. You could then test them by adding a mock contract, kind of how we had one for ExchangeOrderHandler